diff --git a/cmake/AddEventLibrary.cmake b/cmake/AddEventLibrary.cmake
index 04f5837e9..835518c43 100644
--- a/cmake/AddEventLibrary.cmake
+++ b/cmake/AddEventLibrary.cmake
@@ -102,6 +102,7 @@ macro(add_event_library LIB_NAME)
         set(LIB_OUTER_INCLUDES NONE)
     endif()
     set(ADD_EVENT_LIBRARY_INTERFACE)
+    set(INNER_LIBRARIES)
 
     if (${EVENT_LIBRARY_STATIC})
         add_library("${LIB_NAME}_static" STATIC ${LIB_SOURCES})
